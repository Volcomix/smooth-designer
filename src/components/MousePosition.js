//@flow
import React, { type Node } from 'react'

export type Props = {
  children?: Node,
  className?: string,
  isActive: boolean,
  onMouseMove: (x: number, y: number) => void,
}

class MousePosition extends React.Component<Props> {
  container: ?HTMLDivElement

  render() {
    const { children, className } = this.props
    return (
      <div
        className={className}
        onMouseMove={this.handleMouseMove}
        ref={container => (this.container = container)}
      >
        {children}
      </div>
    )
  }

  handleMouseMove = ({ clientX, clientY }: MouseEvent) => {
    if (this.props.isActive && this.container) {
      const { width, height } = this.container.getBoundingClientRect()
      this.props.onMouseMove(clientX - width / 2, clientY - height / 2)
    }
  }
}

export default MousePosition
