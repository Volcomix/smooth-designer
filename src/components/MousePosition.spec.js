//@flow
import React from 'react'
import { shallow, mount } from 'enzyme'
import MousePosition, { type Props } from './MousePosition'

const defaultProps = (): Props => ({
  isActive: true,
  onMouseMove: jest.fn(),
  onMouseUp: jest.fn(),
})

const setup = setupProps => {
  const props: Props = { ...defaultProps(), ...setupProps }
  const wrapper = mount(<MousePosition {...props} />)
  const instance: MousePosition = wrapper.instance()
  if (instance.container) {
    jest
      .spyOn(instance.container, 'getBoundingClientRect')
      .mockImplementation(() => ({ width: 800, height: 600 }))
  }
  return { wrapper, props }
}

it('renders without crashing', () => {
  const wrapper = shallow(
    <MousePosition {...defaultProps()} className="MousePosition">
      <div id="content" />
    </MousePosition>,
  )
  expect(wrapper).toMatchSnapshot()
})

describe('if isActive is true', () => {
  it('calls onMouseMove when the mouse moves', () => {
    const { wrapper, props } = setup()
    wrapper.simulate('mouseMove', { clientX: 410, clientY: 320 })
    expect(props.onMouseMove).toHaveBeenCalledWith(10, 20)
  })

  it('calls onMouseUp when the mouse button is released', () => {
    const { wrapper, props } = setup()
    wrapper.simulate('mouseUp')
    expect(props.onMouseUp).toHaveBeenCalled()
  })
})

describe('if isActive is false', () => {
  it('does not call onMouseMove when the mouse moves', () => {
    const { wrapper, props } = setup({ isActive: false })
    wrapper.simulate('mouseMove', { clientX: 410, clientY: 320 })
    expect(props.onMouseMove).not.toHaveBeenCalled()
  })

  it('does not call onMouseUp when the mouse button is release', () => {
    const { wrapper, props } = setup({ isActive: false })
    wrapper.simulate('mouseUp')
    expect(props.onMouseUp).not.toHaveBeenCalled()
  })
})
