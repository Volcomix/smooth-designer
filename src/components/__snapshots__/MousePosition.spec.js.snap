// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders without crashing 1`] = `
<div
  className="MousePosition"
  onMouseMove={[Function]}
  onMouseUp={[Function]}
>
  <div
    id="content"
  />
</div>
`;
